README
*** README for triple three (3.3.3 :) ***

Read the DOCS/diffs file for update information.

Welcome to NUTS if you're a first time user. Hopefully you'll find it a
simple system to use thanks to the helpfiles and the documentation , if not
then please feel free to email me at the address at the end of this file.
Before I continue though I must stress that I WILL NOT SORT OUT BUGS OF YOUR
OWN MAKING OR WRITE CODE FOR YOU! If you write your own code or alter my
code and fuck it up thats your problem, I have my own projects to do and I
am *not* interested in sorting out yours. I've had one too many emails from
people asking me to sort out the mess they've made and I'm getting rather sick 
of it. However, despite me putting this request in the 3.3.1 docs some people 
have seen fit to ignore it so heres an added incentive: mail me about your cock
ups and not only will I not reply to the mail but I will keep a note of your 
email address and I will not reply to *any* mails you send me even if its about
problems with *my* code. Its your choice. 

I know I've said this before but this time I really mean it. This is the final
version of NUTS 3. I have ideas in mind for a version 4 but thats years away
so in the meantime this is it :) And to prove I mean it this is the first
version of NUTS for a long long time without a backdoor in it as I'm sure 101
hackers will test and find out to be true :)

Unfortunately I still haven't found out what the colour reset code for
SCO consoles and strict ansi terminals is so if you're using one of these 
things may look messy. No one mailed me the code and the SCO terminfo 
translation tools proved less than helpful so theres not much I can do. Sorry.

Now go and read the DOCS/compiling file for info on how to get started if 
you're new to NUTS and all the other files there will be usefull in time also.

Neil Robertson - November 1996

email: neil@ogham.demon.co.uk
home page     : http://www.wso.co.uk/neil.html
NUTS home page: http://www.wso.co.uk/nuts.html

NB: If anyone is also interested in MUDs I've written a MUD system called 
    Ogham - ftp.ccs.neu.edu:/pub/mud/servers/misc/ogham/ogham262.tar.gz
    Its not as complex as perhaps LPmuds or Dikumuds are but if you know C and 
    want to set up a low to medium complexity mud then this is just what you 
    need :) 
    


docs/changelog
ChangeLog
-=-=-=-=-

Look for the variable "talkername" in bolts.h and change this to the name 
of your talker.

Added a famous quote feature whenever you login.

Increased the buffer size from 15 to 25 lines so that people can write
long profiles and messages if they feel so inclined.

Changed the command prompt from COM> to something more suitable:
Fred@Money Mansion:/home/entrance_lobby>

Prompt is in the format of:
You@Talker name:/home_or_name_of_remote_talker/room_you_are_in>  

Added a space to the emote code so that you won't be emoting like this:

Fred@Money Mansion:/home/entrance_lobby> ;flies around.
Fredflies around.

Added the ability to .tell and .pemote to anyone who is in a netlink.

Modified the .invis ability so that only users your level and higher know that you are 
online. People can .tell, .pemote, .beep, .wake, .cprog, .join (etc.) you if they know you 
are online.

Also, you are able to log on and off invis.  If you are invis, only Admins
and higher can see who you are via the "SIGN ON:" message.

Changed the system messages of .clone and .dest to something more suitable.
Somehow I don't think whispering sharp spells fits into a talker's theme.

Changed the .mode prompt from COM> to Money Mansion> and if you are invis,
it will say Money Mansion (Invis)>

Users with swear words in their name cannot login.

Added a prelogin feature where Wizards and above can see who is logging in.

Added some code at the end of the logout function which tells the user
what site they are logging in from and for how long they were logged in.

Added a few lines in the help_commands function to tell you the number
of commands.

Moved the .semote command to .emoteall and gave it to JrWizzes instead.
Moved the .wizshout command to .wiz

Changed the shortcut keys:

> = .tell
< = .pemote
; = .emote
- = .echo

Erased the NEW level.  If you want it, copy it from Neil's NUTS 3.3.3 code.

Removed helpfiles/mainhelp. Now when you issue the .help command, it will 
tell you the commands available. Less muss, less fuss...

Changed the names of the levels from USER,WIZ,ARCH,GOD,* to
Dunce,User,JrWiz,Wizard,Honor,Admin,God,Creator.  Creator is an extra
level above God and is used for debugging purposes and to demote Gods.  Honor 
is an Admin/God from another talker.  Dunce is for consequencing users if they
break the rules, Dunces can only do .help, .look, .who and .quit.  JrWiz is
a Wizard in training.

Also you can't .delete,.promote,.demote,.ban,.unban,.reboot or .shutdown
over a netlink because of certain security risks involved.  That was a bug
by Neil that I corrected.

07/2002 - Changed the output of 'who' so that people of a level lower than you cannot see 
what room you are in and bug you.

Added the following commands:

User level:

.act - a social command, see .help act for details.

.emergency - an emergency channel to wizards in case the user gets into
	     problems.  The channel is one way only.

.wizzes - gives you a list of Wizards and higher.
For every Wizard you promote, you need to add them to datafiles/wizzes

.staff - tells you a list of the VISIBLE wizards on.

.rules - gives the user a list of rules.
Rules are in the file datafiles/rules

.join - Will take you to the specified person, wherever they are.
        Like if the user's in the corridor and you're elsewhere, they can
        .join Fred for example to get to you.  You can't join someone if
	they're in the wizroom.

.lexamine, .ltell, .lwho will perform .examine, .tell and .who on the home
site if you're netlinked.

.chat - Is a handy dandy chat channel I made loosely based on .shout.
        Usage: .chat <your message>
	.chat on <to turn chat on>
	.chat off <to turn it off>

.uptime - Will show you how long the talker has been up and running.

.time - Will show you the system time.

.roulette - Russian roulette

05/17/01 - Changed the default description of a new user from "a Newcomer" 
to "hasn't used .desc yet"

Wizard level:

Made Wizards able to .clone, .dest, etc.
Changed wizshouts to look like this: 
(Wiz) Fred: wibble

Added the on and off parameters to .wiz so you can turn them on/off
with .wiz on & .wiz off respectively.

.arrest - will arrest someone and toss them in jail if they're bad &
	  keep them in jail even if they .quit...when they are arrested,
	  only .quit and .rules work.
.unarrest - will release them from jail
.beep - will beep someone
.echoall - will echo something to all rooms
.wizemote - will emote something to all wiz
.lpeople - runs .people on the home site if you're netlinked
.lbcast - .bcasts something on the ohme site if you're netlinked.
.bring - brings a user to you.
.force - force anyone below your level to do anything.

Honor level:

.honor - it's an Honor channel
	Usage:  .honor <message>
		.honor on <turns it on>
		.honor off <turns it off>
.honoremote - You can emote on the channel using this.

Admin level:

.admin - it's an Admin channel
        Usage:  .admin <message>         
                .admin on <turns it on>  
                .admin off <turns it off>
.adminemote - Allows you to emote on the channel.

God level:

.god - it's a God channel
        Usage:  .god <message>
                .god on <turns it on>
                .god off <turns it off>
.godemote - You can emote on the channel using this.

That's about it.

Please email me with any suggestions, complaints, gripes, etc.

Nathan D Richards <nathanr@nathanr.com>
Aka Sunspot on Paradise (paradise.nathanr.com 7000), the home of Bolts.


bolts.h
/****************** Header file for Bolts version 2.0 ******************/

#define DATAFILES "datafiles"
#define USERFILES "userfiles"
#define HELPFILES "helpfiles"
#define TEMPFILES "tempfiles"
#define HANGDICT  "hangman_words"
#define ALERTFILE "alert"
#define MAILSPOOL "mailspool"
#define CONFIGFILE "config"
#define NEWSFILE "newsfile"
#define MAPFILE "mapfile"
#define WIZZESFILE "wizzes"
#define RULEFILE "rules"
#define SITEBAN "siteban"
#define USERBAN "userban"
#define SYSLOG "syslog"
#define MOTD1 "motd1"
#define MOTD2 "motd2"

#define OUT_BUFF_SIZE 1000
#define MAX_WORDS 10
#define WORD_LEN 40
#define ARR_SIZE 1000
#define MAX_LINES 25
#define NUM_COLS 21

#define USER_NAME_LEN 12
#define USER_DESC_LEN 30
#define AFK_MESG_LEN 60
#define PHRASE_LEN 40
#define PASS_LEN 20 /* only the 1st 8 chars will be used by crypt() though */
#define BUFSIZE 1000
#define ROOM_NAME_LEN 20
#define ROOM_LABEL_LEN 5
#define ROOM_DESC_LEN 810 /* 10 lines of 80 chars each + 10 nl */
#define TOPIC_LEN 60
#define MAX_LINKS 10
#define SERV_NAME_LEN 80
#define SITE_NAME_LEN 80
#define VERIFY_LEN 20
#define REVIEW_LINES 15
#define REVTELL_LINES 5
#define REVIEW_LEN 200
/* DNL (Date Number Length) will have to become 12 on Sun Sep 9 02:46:40 2001 
   when all the unix timers will flip to 1000000000 :) */
#define DNL 11 
#define ICQ_LEN 10
#define URL_LEN 50
#define AGE_LEN 2
#define EMAIL_LEN 50
#define USER_MIN_LEN 3

#define PUBLIC 0
#define PRIVATE 1
#define FIXED 2
#define FIXED_PUBLIC 2
#define FIXED_PRIVATE 3

#define DUNCE 0
#define USER 1
#define JRWIZ 2
#define WIZARD 3
#define HONORED 4
#define ARCH 5
#define SU 6

#define USER_TYPE 0
#define CLONE_TYPE 1
#define REMOTE_TYPE 2
#define CLONE_HEAR_NOTHING 0
#define CLONE_HEAR_SWEARS 1
#define CLONE_HEAR_ALL 2

/* The elements vis, ignall, prompt, command_mode etc could all be bits in 
   one flag variable as they're only ever 0 or 1, but I tried it and it
   made the code unreadable. Better to waste a few bytes */
struct user_struct {
	char name[USER_NAME_LEN+1];
	char desc[USER_DESC_LEN+1];
	char pass[PASS_LEN+6];
	char in_phrase[PHRASE_LEN+1],out_phrase[PHRASE_LEN+1];
	char buff[BUFSIZE],site[81],last_site[81],page_file[81];
	char mail_to[WORD_LEN+1],revbuff[REVTELL_LINES][REVIEW_LEN+2];
	char afk_mesg[AFK_MESG_LEN+1],inpstr_old[REVIEW_LEN+1];
	struct room_struct *room,*invite_room;
	int type,port,site_port,login,socket,attempts,buffpos,filepos;
	int vis,ignall,prompt,command_mode,muzzled,charmode_echo; 
	int level,misc_op,remote_com,edit_line,charcnt,warned;
	int last_login_len,ignall_store,clone_hear,afk,arrested;
	int edit_op,colour,ignshout,igntell,ignchat,ignwiz,ignhonor,ignadmin,igngod,revline;
	time_t last_input,last_login,total_login,read_mail;
	char *malloc_start,*malloc_end;
        char icq[ICQ_LEN+1],url[URL_LEN+1],email[EMAIL_LEN+1];
        int age,gender,hidemail,autofwd;
	int first,twin,tlose,tdraw;
	char array[10];
	/* Hangman Variables */
	char hang_word[WORD_LEN],hang_word_show[WORD_LEN],hang_guess[WORD_LEN];
	int hang_stage;
        /* Bank Account Info */
        int bank_balance,bank_update,bank_temp;
	int editing;
	struct netlink_struct *netlink,*pot_netlink;
	struct user_struct *prev,*next,*owner,*opponent;
	struct blackjack_game_struct *bj_game;
	struct puzzle_struct *puzzle;  /* Puzzle field  */
	};

typedef struct user_struct* UR_OBJECT;
UR_OBJECT user_first,user_last;

struct room_struct {
	char name[ROOM_NAME_LEN+1];
	char label[ROOM_LABEL_LEN+1];
	char desc[ROOM_DESC_LEN+1];
	char topic[TOPIC_LEN+1];
	char revbuff[REVIEW_LINES][REVIEW_LEN+2];
	int inlink; /* 1 if room accepts incoming net links */
	int access; /* public , private etc */
	int revline; /* line number for review */
	int mesg_cnt;
	char netlink_name[SERV_NAME_LEN+1]; /* temp store for config parse */
	char link_label[MAX_LINKS][ROOM_LABEL_LEN+1]; /* temp store for parse */
	struct netlink_struct *netlink; /* for net links, 1 per room */
	struct room_struct *link[MAX_LINKS];
	struct room_struct *next;
	};

typedef struct room_struct *RM_OBJECT;
RM_OBJECT room_first,room_last,jail;
RM_OBJECT create_room();

/* Netlink stuff */
#define UNCONNECTED 0 
#define INCOMING 1 
#define OUTGOING 2
#define DOWN 0
#define VERIFYING 1
#define UP 2
#define ALL 0
#define IN 1
#define OUT 2

/* Structure for net links, ie server initiates them */
struct netlink_struct {
	char service[SERV_NAME_LEN+1];
	char site[SITE_NAME_LEN+1];
	char verification[VERIFY_LEN+1];
	char buffer[ARR_SIZE*2];
	char mail_to[WORD_LEN+1];
	char mail_from[WORD_LEN+1];
	FILE *mailfile;
	time_t last_recvd; 
	int port,socket,type,connected;
	int stage,lastcom,allow,warned,keepalive_cnt;
	int ver_major,ver_minor,ver_patch;
	struct user_struct *mesg_user;
	struct room_struct *connect_room;
	struct netlink_struct *prev,*next;
	};

typedef struct netlink_struct *NL_OBJECT;
NL_OBJECT nl_first,nl_last;
NL_OBJECT create_netlink();

char *syserror="Sorry, a system error has occured";
char *nosuchroom="There is no such room.\n";
char *nosuchuser="There is no such user or remote user does not have a local account.\n";
char *notloggedon="There is no one of that name logged on.\n";
char *invisenter="Someone enters the room.\n";
char *invisleave="Someone leaves the room.\n";
char *invisname="Someone";
char *noswearing="Hey potty mouth!  No swearing.\n";

/* Variables that create a new look conveniently stored in one place! */
char *talkername="Paradise";
char *talkerowner="Sunspot";
char *boltsversion="2.0";
char *talkeraddr="paradise.nathanr.com 7000";
/* you can change this for whatever sig you want - of just "" if you don't want
   to have a sig file attached at the end of emails */
char *talker_signature=
"+--------------------------------------------------------------------------+\n\
| This message has been smailed to you on The Paradise Talker, and this is |\n\
|      your auto-forward.  Please do not reply directly to this email.     |\n\
|                                                                          |\n\
|          Paradise - A talker running at paradise.nathanr.com 7000        |\n\
|      email 'paradise@nathanr.com' if you have any questions/comments     |\n\
+--------------------------------------------------------------------------+\n";

char *level_name[]={
"Dunce","User","JrWiz","Wizard","Honor","Admin","God","Creator"
};

char *command[]={
"quit",    "look",     "mode",      "say",    "shout",
"tell",    "emote",    "emoteall",  "pemote", "echo",
"go",      "ignall",   "prompt",    "desc",   "inphr",
"outphr",  "public",   "private",   "letmein","invite",
"topic",   "move",     "bcast",     "who",    "people",
"help",    "shutdown", "news",      "read",   "write",
"wipe",    "search",   "review",    "home",   "status",
"version", "rmail",    "smail",     "dmail",  "from",
"entpro",  "examine",  "rmst",      "rmsn",   "netstat",
"netdata", "connect",  "disconnect","passwd", "kill",
"promote", "demote",   "listbans",  "ban",    "unban",
"vis",     "invis",    "site",      "wake",   "wiz",
"muzzle",  "unmuzzle", "map",       "logging",
"system",  "charecho", "clearline", "fix",    "unfix",
"viewlog", "revclr",   "clone",     "destroy",
"myclones","allclones","switch",    "csay",   "chear",
"rstat",   "swban",    "afk",       "cls",    "colour",
"ignshout","igntell",  "suicide",   "delete", "reboot",
"recount", "revtell",  "think",     "mumble",
"sing",    "wizzes",   "rules",     "beep",
"cprog",
"wizemote",
"join",    "chat",     "chatemote", 
"lexamine","ltell",    "lwho",      "lpeople","lbcast",
"admin",   "adminemote","emergency","act",
"arrest",   "unarrest",  "honor","honoremote",
"god",      "godemote", "echoall","echoto",
"staff",    "bring",    "time", "roulette",
"sitechange", "sayto",  "set",  "squit", "force",
"makeinvis", "makevis", "finger", "nslookup", "clsall", 
"calendar", "fortune",  "cemote", "icqpage", "traceroute", 
"ping",     "tictac",   "hangman", "guess",  "givecash",
"lendcash",  "balance", "greet",   "bjack",  "alert",
"modify",    "save",    "bbcast",  "fmail",  "show",
"ctopic",    "bfrom",   "mutter",  "listafks", "puzzle",
"ignchat",   "ignwiz",  "ignhonor","ignadmin","igngod",
"ranks",     "shift",   "sshout",  "rename",  "settime", 
"srboot",    "urename", "*"
};


/* Values of commands , used in switch in exec_com() */
enum comvals {
QUIT,     LOOK,     MODE,     SAY,    SHOUT,
TELL,     EMOTE,    EMOTEALL, PEMOTE, ECHO,
GO,       IGNALL,   PROMPT,   DESC,   INPHRASE,
OUTPHRASE,PUBCOM,   PRIVCOM,  LETMEIN,INVITE,
TOPIC,    MOVE,     BCAST,    WHO,    PEOPLE,
HELP,     SHUTDOWN, NEWS,     READ,   WRITE,
WIPE,     SEARCH,   REVIEW,   HOME,   STATUS,
VER,      RMAIL,    SMAIL,    DMAIL,  FROM,
ENTPRO,   EXAMINE,  RMST,     RMSN,   NETSTAT,
NETDATA,  CONN,     DISCONN,  PASSWD, KILL,
PROMOTE,  DEMOTE,   LISTBANS, BAN,    UNBAN,
VIS,      INVIS,    SITE,     WAKE,   WIZ,
MUZZLE,   UNMUZZLE, MAP,      LOGGING,
SYSTEM,   CHARECHO, CLEARLINE,FIX,    UNFIX,
VIEWLOG,  REVCLR,   CREATE,   DESTROY,
MYCLONES, ALLCLONES,SWITCH,   CSAY,   CHEAR,
RSTAT,    SWBAN,    AFK,      CLS,    COLOUR,
IGNSHOUT, IGNTELL,  SUICIDE,  DELETE, REBOOT,
RECOUNT,  REVTELL,  THINK,    MUMBLE,
SING,     WIZZES,   RULES,    BEEP,
CPROG,
WIZEMOTE,
JOIN,     CHAT,     CHATEMOTE, 
LEXAMINE, LTELL,    LWHO,     LPEOPLE,LBCAST,
ADMIN,    ADMINEMOTE,EMERGENCY,ACT,
ARREST,   UNARREST, HONOR,  HONOREMOTE,
GOD,      GODEMOTE, ECHOALL,ECHOTO,
STAFF ,   BRING,    TIME,   ROULETTE,
SITECHANGE, SAYTO,  SET,    SQUIT, FORCE,
MAKEINVIS, MAKEVIS, FINGER, NSLOOKUP, CLSALL,
CALENDAR, FORTUNE,  CEMOTE, ICQPAGE, TRACEROUTE,
PING,     TICTAC,   HANGMAN, GUESS,  GIVECASH,
LENDCASH, BALANCE,  GREET,  BJACK,   ALERT,
MODIFY,   SAVE,     BBCAST, FMAIL,   SHOW,
CTOPIC,   BFROM,    MUTTER, LISTAFKS,PUZZLE,
IGNCHAT,  IGNWIZ,   IGNHONOR,IGNADMIN,IGNGOD,
RANKS,    SHIFT,    SSHOUT,  RENAME,SETTIME,
SRBOOT,   URENAME
} com_num;


/* These are the minimum levels at which the commands can be executed. 
   Alter to suit. */
int com_level[]={
DUNCE, DUNCE, USER, USER, USER,
USER,USER,WIZARD,USER,USER,
USER,USER,USER, USER,USER,
USER,USER,USER,USER,USER,
USER,JRWIZ, JRWIZ ,DUNCE, JRWIZ,
DUNCE, SU, USER,USER, USER,
WIZARD, USER,USER,USER, USER,
USER, USER, USER,USER,USER,
USER,USER,USER, USER, WIZARD,
ARCH,HONORED, HONORED, USER,WIZARD,
ARCH,ARCH,WIZARD, ARCH,ARCH,
WIZARD,WIZARD,WIZARD, WIZARD,JRWIZ,
JRWIZ, WIZARD, USER,SU,
WIZARD, USER, ARCH,ARCH, ARCH,
WIZARD ,USER,JRWIZ,JRWIZ,
JRWIZ,JRWIZ,JRWIZ,JRWIZ,JRWIZ,
WIZARD, ARCH,USER,USER ,USER,
USER,USER,USER, ARCH, SU,
ARCH, USER,USER,USER,
USER,USER,USER,WIZARD,
USER,
JRWIZ,
USER,USER,USER,
USER,USER,USER,JRWIZ,JRWIZ,
ARCH,ARCH,USER,USER,
JRWIZ,JRWIZ,HONORED,HONORED,
SU,SU,JRWIZ,WIZARD,
ARCH,JRWIZ,USER,USER,
ARCH,USER,USER,SU,WIZARD,
ARCH,ARCH,ARCH,USER,ARCH,
USER,USER,JRWIZ,USER,ARCH,
USER,USER,USER,USER,ARCH,
USER,USER,USER,USER,USER,
SU,ARCH,ARCH,USER,ARCH,
ARCH,USER,USER,WIZARD,USER,
USER,JRWIZ,HONORED,ARCH,SU,
USER,USER,USER,ARCH,ARCH,
ARCH,ARCH
};

/* 
Colcode values equal the following:
RESET,BOLD,BLINK,REVERSE

Foreground & background colours in order..
BLACK,RED,GREEN,YELLOW/ORANGE,
BLUE,MAGENTA,TURQUIOSE,WHITE
*/

char *colcode[NUM_COLS]={
/* Standard stuff */
"\033[0m", "\033[1m", "\033[4m", "\033[5m", "\033[7m",
/* Foreground colour */
"\033[30m","\033[31m","\033[32m","\033[33m",
"\033[34m","\033[35m","\033[36m","\033[37m",
/* Background colour */
"\033[40m","\033[41m","\033[42m","\033[43m",
"\033[44m","\033[45m","\033[46m","\033[47m"
};

/* Codes used in a string to produce the colours when prepended with a '~' */
char *colcom[NUM_COLS]={
"RS","OL","UL","LI","RV",
"FK","FR","FG","FY",
"FB","FM","FT","FW",
"BK","BR","BG","BY",
"BB","BM","BT","BW"
};


char *month[12]={
"January","February","March","April","May","June",
"July","August","September","October","November","December"
};

char *day[7]={
"Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"
};

char *noyes1[]={ " NO","YES" };
char *noyes2[]={ "NO ","YES" };
char *offon[]={ "OFF","ON " };

/* gender strings */

char *genders[] = {"neuter", "male", "female"};
char *heshe[] = {"it","he","she"};
char *himher[] = {"it","him","her"};
char *hisher[] = {"its","his","her"};
char *capheshe[] ={"It","He","She"};

/* level abbreviations */
char *levels[] = {"D", "U", "J", "W", "H", "A", "G", "C"};

/* Big letter array map - Used by greet() */
/* This was added by Andy Collington, Creator of Amnuts */

int biglet[26][5][5] = {
    0,1,1,1,0,1,0,0,0,1,1,1,1,1,1,1,0,0,0,1,1,0,0,0,1,
    1,1,1,1,0,1,0,0,0,1,1,1,1,1,0,1,0,0,0,1,1,1,1,1,0,
    0,1,1,1,1,1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,1,1,1,
    1,1,1,1,0,1,0,0,0,1,1,0,0,0,1,1,0,0,0,1,1,1,1,1,0,
    1,1,1,1,1,1,0,0,0,0,1,1,1,1,0,1,0,0,0,0,1,1,1,1,1,
    1,1,1,1,1,1,0,0,0,0,1,1,1,1,0,1,0,0,0,0,1,0,0,0,0,
    0,1,1,1,0,1,0,0,0,0,1,0,1,1,0,1,0,0,0,1,0,1,1,1,0,
    1,0,0,0,1,1,0,0,0,1,1,1,1,1,1,1,0,0,0,1,1,0,0,0,1,
    0,1,1,1,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,1,1,1,0,
    0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,1,0,0,0,1,0,1,1,1,0,
    1,0,0,0,1,1,0,0,1,0,1,0,1,0,0,1,0,0,1,0,1,0,0,0,1,
    1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,1,1,1,1,
    1,0,0,0,1,1,1,0,1,1,1,0,1,0,1,1,0,0,0,1,1,0,0,0,1,
    1,0,0,0,1,1,1,0,0,1,1,0,1,0,1,1,0,0,1,1,1,0,0,0,1,
    0,1,1,1,0,1,0,0,0,1,1,0,0,0,1,1,0,0,0,1,0,1,1,1,0,
    1,1,1,1,0,1,0,0,0,1,1,1,1,1,0,1,0,0,0,0,1,0,0,0,0,
    0,1,1,1,0,1,0,0,0,1,1,0,1,0,1,1,0,0,1,1,0,1,1,1,0,
    1,1,1,1,0,1,0,0,0,1,1,1,1,1,0,1,0,0,1,0,1,0,0,0,1,
    0,1,1,1,1,1,0,0,0,0,0,1,1,1,0,0,0,0,0,1,1,1,1,1,0,
    1,1,1,1,1,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,
    1,0,0,0,1,1,0,0,0,1,1,0,0,0,1,1,0,0,0,1,1,1,1,1,1,
    1,0,0,0,1,1,0,0,0,1,0,1,0,1,0,0,1,0,1,0,0,0,1,0,0,
    1,0,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,0,1,1,1,0,0,0,1,
    1,0,0,0,1,0,1,0,1,0,0,0,1,0,0,0,1,0,1,0,1,0,0,0,1,
    1,0,0,0,1,0,1,0,1,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,
    1,1,1,1,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,1,1,1,1
    };

/* Symbol array map - Used by greet() */
int bigsym[32][5][5] = {
    0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,
    0,1,0,1,0,0,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,1,0,1,0,1,1,1,1,1,0,1,0,1,0,1,1,1,1,1,0,1,0,1,0,
    0,1,1,1,1,1,0,1,0,0,0,1,1,1,0,0,0,1,0,1,1,1,1,1,0,
    1,1,0,0,1,1,1,0,1,0,0,0,1,0,0,0,1,0,1,1,1,0,0,1,1,
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,1,1,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,1,0,
    0,1,1,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,1,1,0,0,
    1,0,1,0,1,0,1,1,1,0,1,1,1,1,1,0,1,1,1,0,1,0,1,0,1,
    0,0,1,0,0,0,0,1,0,0,1,1,1,1,1,0,0,1,0,0,0,0,1,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,1,0,0,0,
    0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,0,0,0,1,1,0,0,0,
    0,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,0,
    0,1,1,1,0,1,0,0,1,1,1,0,1,0,1,1,1,0,0,1,0,1,1,1,0,
    0,0,1,0,0,0,1,1,0,0,0,0,1,0,0,0,0,1,0,0,0,1,1,1,0,
    1,1,1,1,0,0,0,0,0,1,0,1,1,1,0,1,0,0,0,0,1,1,1,1,1,
    1,1,1,1,0,0,0,0,0,1,0,1,1,1,0,0,0,0,0,1,1,1,1,1,0,
    0,0,1,1,0,0,1,0,0,0,1,0,0,1,0,1,1,1,1,1,0,0,0,1,0,
    1,1,1,1,1,1,0,0,0,0,1,1,1,1,0,0,0,0,0,1,1,1,1,1,0,
    0,1,1,1,0,1,0,0,0,0,1,1,1,1,0,1,0,0,0,1,0,1,1,1,0,
    1,1,1,1,1,0,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,
    0,1,1,1,0,1,0,0,0,1,0,1,1,1,0,1,0,0,0,1,0,1,1,1,0,
    1,1,1,1,1,1,0,0,0,1,1,1,1,1,1,0,0,0,0,1,0,0,0,0,1,
    0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,0,0,
    0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,1,0,
    0,0,0,0,0,1,1,1,1,1,0,0,0,0,0,1,1,1,1,1,0,0,0,0,0,
    0,1,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,
    0,1,1,1,1,0,0,0,0,1,0,0,1,1,1,0,0,0,0,0,0,0,1,0,0,
    0,1,0,0,0,1,0,1,1,1,1,0,1,0,1,1,0,1,1,1,0,1,1,1,0
    };

#define GM_TDIM         4       /* used by puzzle functions */
#define GM_LT           0
#define GM_RG           1
#define GM_FD           2
#define GM_BK           3

/* GAEN Puzzle structure */
struct puzzle_struct {
        int table[GM_TDIM][GM_TDIM];    /* table with numbers */
        int clin,ccol;                  /* current position */
};


/* These MUST be in lower case - the contains_swearing() function converts
   the string to be checked to lower case before it compares it against
   these. Also even if you dont want to ban any words you must keep the 
   star as the first element in the array. */
char *swear_words[]={
"fuck","shit","cunt","*"
};

char verification[SERV_NAME_LEN+1];
char text[ARR_SIZE*2];
char word[MAX_WORDS][WORD_LEN+1];
char wrd[8][81];
char progname[40],confile[40];
time_t boot_time;
jmp_buf jmpvar;

int port[3],listen_sock[3],wizport_level,minlogin_level;
int colour_def,password_echo,ignore_sigterm;
int max_users,max_clones,num_of_users,num_of_logins,heartbeat;
int login_idle_time,user_idle_time,config_line,word_count;
int tyear,tmonth,tday,tmday,twday,thour,tmin,tsec;
int mesg_life,system_logging,prompt_def,no_prompt;
int force_listen,gatecrash_level,min_private_users;
int ignore_mp_level,rem_user_maxlevel,rem_user_deflevel;
int destructed,mesg_check_hour,mesg_check_min,net_idle_time;
int keepalive_interval,auto_connect,ban_swearing,crash_action;
int time_out_afks,allow_caps_in_name,rs_countdown;
int charecho_def,time_out_maxlevel;
time_t rs_announce,rs_which;
UR_OBJECT rs_user;

extern __const char *__const sys_errlist[];

char *long_date();
